subprojects/core/src/main/java/org/gradle/initialization/ProjectSpec.java

```diff
@@ -16,20 +16,18 @@
 package org.gradle.initialization;
 
 import org.gradle.api.InvalidUserDataException;
-import org.gradle.api.internal.project.ProjectIdentifier;
-import org.gradle.api.internal.project.ProjectRegistry;
 
 public interface ProjectSpec {
     /**
      * Determines whether the given registry contains at least 1 project which meets this spec.
      */
-    boolean containsProject(ProjectRegistry<? extends ProjectIdentifier> registry);
+    boolean containsProject(ProjectDescriptorRegistry registry);
 
     /**
      * Returns the single project in the given registry which meets this spec.
      * @return the project.
      * @throws InvalidUserDataException When project cannot be selected due to some user input mismatch, or when there are no matching projects
      * or multiple matching projects.
      */
-    <T extends ProjectIdentifier> T selectProject(String settingsDescription, ProjectRegistry<? extends T> registry) throws InvalidUserDataException;
+    DefaultProjectDescriptor selectProject(String settingsDescription, ProjectDescriptorRegistry registry) throws InvalidUserDataException;
```

Comment:
💭 I believe we can actually change the return type to be `org.gradle.api.internal.project.ProjectIdentifier`, or even `Path`. I believe the only actual usage of this returned value is here:
https://github.com/gradle/gradle/blob/ea1f76efda73a4789114a338c0b7719fdbdf7ef8/subprojects/core/src/main/java/org/gradle/initialization/InstantiatingBuildLoader.java#L30

And we only ever use the path of the project there.

---

Can you see if we can avoid the usage of the implementation type `DefaultProjectDescriptor` in other places as well?